devlib/utils/ssh.py

```diff
@@ -180,6 +181,7 @@ def __init__(self,
         logger.debug('Logging in {}@{}'.format(username, host))
         timeout = timeout if timeout is not None else self.default_timeout
         self.conn = ssh_get_shell(host, username, password, self.keyfile, port, timeout, False, None)
+        atexit.register(self.close)
```

Comment:
Should be implemented inside `__del__` instead, i.e. close the connection when the object is destroyed, not when the interpreter exists.

Reply:
When implementing with `__del__` the method was not called unless the program exited cleanly and still appeared to leave the the connections open.

Reply:
OK, I was wrong on this. The docs for `__del__` state that it's not guaranteed to be called for objects that exist on interpreter exit: https://docs.python.org/3/reference/datamodel.html#object.__del__

I withdraw my previous comment.